Store pod IP in OVS external-ids and use that on teardown

danwinship opened the pull request:
If the pod is torn down after its veth has already been destroyed, then OVS will report "-1" for its ofport, so we can't rely on using that to find the pod IP from the flows.

One way to reproduce this is to "systemctl restart docker" when there's a pod running. The pod teardown attempt will fail because ovscontroller will try to run "ovs-ofctl dump-flows br0 in_port=-1", and then it will leave behind stray flows related to the old IP address (while reporting success to the caller).

This is a new bug in 3.9, and it's probably not related to #12544

rajatchopra commented on pkg/util/ovs/parse.go:
```diff
@@ -361,3 +361,35 @@ func fieldMatches(val, match string, ptype ParseType) bool {
+	return ids, nil
+}
+
+func UnparseExternalIDs(externalIDs map[string]string) string {
```
You could have said 'encode' :)

rajatchopra commented on pkg/util/ovs/parse.go:
```diff
@@ -361,3 +361,35 @@ func fieldMatches(val, match string, ptype ParseType) bool {
+	if externalIDs[0] == '{' && externalIDs[len(externalIDs)-1] == '}' {
+		externalIDs = externalIDs[1:len(externalIDs)-1]
+	}
+	for _, id := range strings.Split(externalIDs, ",") {
```
would strings.Fields or strings.FieldsFunc make it better?
Split/Trim combos are delicate.

dcbw replied:
I kinda want unit tests for the Parse/Unparse functions too.

pravisankar replied:
Yes, considering this is common for find, list and set ops, unit tests will be helpful.

rajatchopra commented on pkg/network/node/ovscontroller.go:
```diff
@@ -223,8 +223,10 @@ func (oc *ovsController) NewTransaction() ovs.Transaction {
-	return oc.ovs.AddPort(hostVeth, -1, "external-ids=sandbox="+sandboxID)
+func (oc *ovsController) ensureOvsPort(hostVeth, sandboxID, podIP string) (int, error) {
+	return oc.ovs.AddPort(hostVeth, -1,
+		fmt.Sprintf(`external-ids=sandbox="%s",ip="%s"`, sandboxID, podIP),
```
Don't remember top off the head. Double check if this is 'external_ids=' or 'external_ids:', pretty sure its the colon in some usage, maybe not while creating the port.

dcbw replied:
@rajatchopra IIRC it's = for setting, but : for finding (eg : is sub-match so you can find an entry by one external-id, while = is match all)

imcsk8 replied:
+1

danwinship replied:
It's not "set" vs "find" though, it's "value of the whole map" vs "value of a single element".

`set external-ids=foo=bar` means `external_ids = map[string]string{ "foo": "bar" }`
`set external-ids:foo=bar` means `external_ids["foo"] = "bar"` (other members are unchanged)
`find external-ids=foo=bar` means `if reflect.DeepEqual(external_ids, map[string]string{ "foo": "bar" }) { ...`
`find external-ids:foo=bar` means `if external_ids["foo"] == "bar" { ...`

pravisankar commented on pkg/network/node/ovscontroller.go:
```diff
@@ -333,35 +335,32 @@ func (oc *ovsController) getPodDetailsBySandboxID(sandboxID string) (int, net.IP
-		return 0, nil, fmt.Errorf("found multiple ofports for sandbox ID %q: %#v", sandboxID, strports)
 	}
-	ofport, err := strconv.Atoi(strports[0])
+	strIDs, err := oc.ovs.Find("interface", "external-ids", "external-ids:sandbox="+sandboxID)
```
getPodDetailsBySandboxID() is called by both UpdatePod() and TeardownPod() which is expected to be called frequently in some cases (e.g: online). If we change ovs.Find() interface to take multiple columns, then we could get strports and strIDs in one call.

danwinship replied:
I'm not sure it's *that* frequent, relative to how expensive the operation is ("not very"). All the other callers of ovs.Find want to use a single column, so it would be a little annoying to force them to pass an array of columns names and then get back an array of arrays as the answer...

(Note that we're switching from "1 ovs-vsctl call plus 1 ovs-ofctl call" to "2 ovs-vsctl calls", so we're not making things worse, just failing to make them better.)

pravisankar commented on pkg/util/ovs/ovs.go:
```diff
@@ -250,11 +250,19 @@ func (ovsif *ovsExec) Set(table, record string, values ...string) error {
+	// We want "bare" values for strings, but we can't pass --bare to ovs-vsctl because
+	// it breaks more complicated types
+	for i, val := range values {
+		if unquoted, err := strconv.Unquote(val); err == nil {
```
Do we really want to ignore the error here?

danwinship replied:
Yes. If `val` is a quoted string, then we know OVS would have quoted it correctly, so Unquote() won't return an error. So if Unquote() returns an error, that means the value wasn't a quoted string. (Eg, it was a number.)

Originally I had

    if val[0] == '"' && val[len(val)-1] == '"' {

before the Unquote call, but then that panics if val is empty (which it can be). And then while fixing it I realized that Unquote was going to do that check anyway so I should just let it do it.

pravisankar commented on pkg/util/ovs/parse.go:
```diff
@@ -361,3 +361,35 @@ func fieldMatches(val, match string, ptype ParseType) bool {
+		}
+		key := parsed[0]
+		value := parsed[1]
+		if value[0] == '"' {
```
ovs.Find() already unquotes the values, is this for list or set op?

danwinship replied:
ovs.Find() unquotes the value if the *whole value* was quoted (meaning, it's a string-value column). But here we're parsing a map-valued column, so ovs.Find won't have removed the internal quotes in the value.

imcsk8 commented:
Code looks good

danwinship commented:
Added more unit tests and updated one comment.

I didn't do anything about the `=` vs `:` thing because the commands just work with the ovs-vsctl syntax, and that's already documented in the ovs-vsctl man page (and we don't really document anything else, like what the returned format of the data is, etc)

pravisankar commented:
/lgtm

pravisankar commented:
/retest
